### PR TITLE
Flake8 Round 2: We have to go deeper

### PR DESCRIPTION
**Pull request by cthoyt**
As a follow-up to #156, this one enables even more flake8 checks from the `flake8-docstrings` plugin. I did a few of the tiny ones, but now the meatiest part is to address `D103` - all public functions need a docstring. I'm going to conscript @hrshdhgd for this bit

Keep in mind all of this flake8/mypy/code style stuff is a bit of a pain in the short term, but it will keep us sane as developers in the future and support the longevity of a community that might come to rely on this package at some point

81mPquqrVLL _AC_SL1000_

### CODE REVIEW

File: sssom/cliques.py [hunk 1]
```diff
@@ -59,7 +59,17 @@ def to_networkx(msdf: MappingSetDataFrame) -> nx.DiGraph:
     return g
 
 
-def split_into_cliques(msdf: MappingSetDataFrame):
+def split_into_cliques(msdf: MappingSetDataFrame) -> List[MappingSetDocument]:
+    """Split MappingSetDataFrames into cliques.
+
+    :param msdf: MappingSetDataFrame object
+    :type msdf: MappingSetDataFrame
```
Comment (cthoyt) [inline sssom/cliques.py:66]: don't add type annotations in the docs. They're already taken care of by type annotations

File: sssom/cliques.py [hunk 2]
```diff
@@ -95,7 +112,16 @@ def invert_dict(d: Dict[str, str]) -> Dict[str, str]:
     return invdict
 
 
-def get_src(src, cid):
+def get_src(src: str, cid: str):
+    """Get source.
```
Comment (cthoyt) [inline sssom/cliques.py:116]: A good docstring doesn't just explain what the function does, but also why it does it

File: sssom/context.py [hunk 3]
```diff
@@ -15,14 +15,29 @@
 
 
 def get_jsonld_context():
+    """Get JSON-LD context.
```
Comment (cthoyt) [inline sssom/context.py:18]: What JSON-LD context does it get? What's it for?

File: sssom/context.py [hunk 4]
```diff
@@ -51,6 +73,11 @@ def add_built_in_prefixes_to_prefix_map(
 
 
 def get_default_metadata() -> Metadata:
+    """Get Default metadata.
```
Comment (cthoyt) [inline sssom/context.py:76]: It's almost never sufficient to just re-write the name of the function. What is default metadata? How was it chosen? Imagine a user could read this docstring and not know what it is or where

Comment (cthoyt) [reply sssom/context.py:76]: Don't be afraid to be redundant of other function's docstrings, since documentation readers might only find the one function they want to know more about.

File: tests/test_data.py [hunk 5]
```diff
@@ -14,22 +14,40 @@
...
+    :param filename: Filename
+    :type filename: TextIO
+    :return: File path
+    :rtype: str
+    """
     return os.path.join(test_data_dir, filename)
 
 
 def ensure_test_dir_exists():
+    """Create directory if not already present."""
```
Comment (cthoyt) [inline tests/test_data.py:29]: Which directory?

Comment (cthoyt) [reply tests/test_data.py:29]: Tbh this shouldn't even be a function - this ensure should be in the module code

File: tests/test_data.py [hunk 6]
```diff
@@ -14,22 +14,40 @@
...
     if not os.path.exists(test_out_dir):
         os.makedirs(test_out_dir)
 
 
 def load_config():
+    """Load configuration.
```
Comment (cthoyt) [inline tests/test_data.py:35]: what configuration? What's in it? What's the configuration for?

File: tests/test_data.py [hunk 7]
```diff
@@ -14,22 +14,40 @@
...
     with open(TEST_CONFIG) as file:
         config = yaml.safe_load(file)
     return config
 
 
 def get_all_test_cases():
+    """Get all test cases.
+
+    :return: List of test cases
+    :rtype: List[SSOMTestCase]
```
Comment (cthoyt) [inline tests/test_data.py:49]: use type annotations so this is checked with mypy (also there's a typo)

File: sssom/cliques.py [hunk 8]
```diff
@@ -59,7 +59,17 @@ def to_networkx(msdf: MappingSetDataFrame) -> nx.DiGraph:
     return g
 
 
-def split_into_cliques(msdf: MappingSetDataFrame):
+def split_into_cliques(msdf: MappingSetDataFrame) -> List[MappingSetDocument]:
+    """Split MappingSetDataFrames into cliques.
```
Comment (cthoyt) [inline sssom/cliques.py:63]: What's a clique?

File: sssom/cliques.py [hunk 9]
```diff
@@ -87,6 +97,13 @@ def split_into_cliques(msdf: MappingSetDataFrame):
 
 
 def invert_dict(d: Dict[str, str]) -> Dict[str, str]:
+    """Invert Dictionary.
```
Comment (cthoyt) [inline sssom/cliques.py:100]: What does it mean to invert a dictionary?

Comment (cthoyt) [reply sssom/cliques.py:100]: What assumptions are necessary before doing this?

File: sssom/context.py [hunk 10]
```diff
@@ -15,14 +15,29 @@
...
+    """Get JSON-LD context.
+
+    :return: JSON-LD context
+    :rtype: Any
+    """
     return json.loads(sssom_external_context, strict=False)
 
 
 def get_built_in_prefix_map() -> PrefixMap:
+    """Get built-in prefix map.
```
Comment (cthoyt) [inline sssom/context.py:36]: What is a prefix map? What's special about the built-in one?

File: tests/test_data.py [hunk 11]
```diff
@@ -14,22 +14,40 @@
 DEFAULT_CONTEXT_PATH = os.path.join(schema_dir, "sssom.context.jsonld")
 
 
-def get_test_file(filename):
+def get_test_file(filename: TextIO) -> str:
+    """Get test file.
+
+    :param filename: Filename
```
Comment (cthoyt) [inline tests/test_data.py:20]: Is TextIO really the right type annotation? Looks like it should be a str or Path

File: sssom/cliques.py (hunk 9 above)
Comment (matentzn) [reply sssom/cliques.py:100]: assumptions:
1. key and value are literals of some kind
2. keyset contains no duplicated, valueset contains no duplicates. Great questions. The code should handle exceptions to these two assumptions.. Can we turn to issue?

Comment (matentzn) [reply sssom/cliques.py:100]: Invert a dictinary means keys become values and values become keys.

File: sssom/cliques.py (hunk 1 above)
Comment (matentzn) [reply sssom/cliques.py:66]: Is this commonly accepted practice? My assumption was that in order for documentation systems like sphinx to generate anything meaningful it would need the type hints in the docstring.

File: sssom/cliques.py (hunk 2 above)
Comment (matentzn) [reply sssom/cliques.py:116]: dont know nothing about clique.py

File: sssom/context.py (hunk 3 above)
Comment (matentzn) [reply sssom/context.py:18]: Can you add these comments as #FIXME to the code instead of this PR? I will add the information whenever I stumble across them :)

File: sssom/context.py (hunk 4 above)
Comment (matentzn) [reply sssom/context.py:76]: Agreed. For this PR to pass to I need to fix all these? Or should fixing these just be on the longer agenda?

**Comment by matentzn** [comment]
@cthoyt Danke danke wie immer!.. :) Do we have to fix all your comments to be able to pass the PR? 

I would like to move to a smaller PR model - not because I don't like big changes (I do), but because we need to address some functional issues now, and these code-quality PRs do conflict a lot with these. Can I trust that:
1) you think your codestyle suggestions are standard and help maintain good quality code
2) your improvement suggestions are necessary to meet the code style requirements?

Thanks for advice!

File: setup.cfg [hunk 12]
```diff
@@ -146,13 +146,10 @@ ignore =
     S318 # don't worry about unsafe xml
     S310 # TODO remove this later and switch to using requests
     # The following are documentation things (remove one at a time in future PRs)
+    BLK100
```
Comment (cthoyt) [inline setup.cfg:149]: You're not allowed to remove code quality checks :p

File: sssom/cliques.py (hunk 1 above)
Comment (cthoyt) [reply sssom/cliques.py:66]: The [`sphinx-autodoc-typehints`](https://github.com/agronholm/sphinx-autodoc-typehints) plugin to Sphinx takes care of this (if that's not already installed on this project then I will add it now). I think it's also obvious that having the type annotations be part of the syntax gives the benefit of it being typo checked

File: tests/test_data.py [hunk 13]
```diff
@@ -37,6 +56,11 @@ def get_all_test_cases():
 
 
 def get_multiple_input_test_cases():
+    """Get multiple input test cases.
+
+    :return: List of test cases
```
Comment (cthoyt) [inline tests/test_data.py:61]: If it's possible to explain what the function does in one line, then you can consider not writing the redundant `:return:`

File: tests/test_data.py (hunk 11 above)
Comment (cthoyt) [reply tests/test_data.py]: Fixed in b47274d

File: tests/test_data.py (hunk 5 above)
Comment (cthoyt) [reply tests/test_data.py]: Fixed in b47274d

File: sssom/cliques.py (hunk 2 above)
Comment (cthoyt) [reply sssom/cliques.py:116]: Sure, but it's definitely possible to read this function and check where it's used then figure out what it does. I would estimate this is 2 minutes effort for me since I have a lot of experience doing this (and know all of the keyboard shortcuts in PyCharm 😛) and maybe it will take 5-10 minutes for @hrshdhgd as he gets more used to doing this kind of digital detective work :)

File: sssom/context.py (hunk 4 above)
Comment (cthoyt) [reply sssom/context.py:76]: Yes. The purpose of this PR is cleanup, so if the job isn't done (and also done well) then there's no advantage to merging it. As you can see, it's indeed possible to game flake8 by writing unhelpful docstrings. It's not really helpful for users to read documentation that just barely passes flake8 - we have to go the extra step and make sure that the docstrings can educate and inform users. As it's the case that we as the developers and maintainers don't currently know what some functions do... well this is not a great situation to be in. It means we need to put in the legwork to figure out what all of the code does (even if we weren't the ones who wrote it).

One alternative is to scale back the scope of this PR (i.e., reinstate the ignore for D103) and delete the _stub_ docstrings, but I think that it's rather the case that @hrshdhgd just needs a bit of training and practice on writing useful docstrings. Tbh after a quick google search, I didn't find any material that really showed what a good docstring was other than the minimum syntactical and semantic requirements for it to be readable by Sphinx, so this will take some effort and a bit more back-and-forth in the comments.

Comment (cthoyt) [reply sssom/context.py:76]: Another note on why this should be done well - it's 99% likely that once the D103 check (this is the one that makes sure functions have docstrings with the correct style) that many of them will never be touched again. If we allow an unhelpful docstring to pass through review, then it will never be fixed, and we'll never have a flake8 error to check it for us, and never another pass flake8 campaign where we organize effort to fix this.

File: sssom/cliques.py (hunk 9 above)
Comment (cthoyt) [reply sssom/cliques.py:100]: Cool @hrshdhgd would you please incorporate this in the docs then

File: sssom/context.py (hunk 3 above)
Comment (cthoyt) [reply sssom/context.py:18]: That's an alternative to the two possibilities I suggested in this comment https://github.com/mapping-commons/sssom-py/pull/158#discussion_r716499976. I would say we should use FIXMEs sparingly as an alternative to fixing the problem now that we found it

**Comment by cthoyt** [comment]
> @cthoyt Danke danke wie immer!.. :) Do we have to fix all your comments to be able to pass the PR?
> 
> I would like to move to a smaller PR model - not because I don't like big changes (I do), but because we need to address some functional issues now, and these code-quality PRs do conflict a lot with these. Can I trust that:
> 
>     1. you think your codestyle suggestions are standard and help maintain good quality code
> 
>     2. your improvement suggestions are necessary to meet the code style requirements?
> 
> 
> Thanks for advice!

I agree, it's annoying to review these large PRs, but there aren't any good ways to partially start requiring certain flake8 checks in certain places that aren't super tricky/hard to set up. I would hope that new functionality doesn't conflict with code-quality PRs - especially this one which is _mostly_ relegated to just adding docstrings (and the occasional code fix up).

> 1. you think your codestyle suggestions are standard and help maintain good quality code

Yes, all of the suggestions I've given are not only for maintaining good quality code, but for the users' benefit. Consider one of the questions I asked about what a function does, and currently nobody can really answer. This is a pretty big liability, and we should figure it out ASAP!

> 2. your improvement suggestions are necessary to meet the code style requirements?

It's hard to say. If the requirements are "pass flake8" then I guess you can get away with doing a thing that makes it pass flake8, but like I mentioned above, this is really not enough. It's hard to write up requirements, but if I had to I would say these are the points you should hit:

1. A good docstring should explain to a user what the function does and why
2. A good docstring explains what the inputs are to the function, and where they might come from
3. A good docstring explains what the output of the function is, if it has one, and where it might get used. If there's no output, then the function probably has a side-effect (like IO) and that should be described
4. A good docstring does all of this using standard documentation style (i.e., the sphinx style in sssom-py) and follows the format suggestions given by flake8 and related code quality assurance scripts (e.g., write the first line of the docstring in the imperative mood)
5. A good docstring gives minimal examples, when possible. Obviously there are many functions that this just isn't really feasible, and this wouldn't make sense. However, documentation goes much further than writing docstrings (disclaimer: the following advice is not in scope of this PR) and if you can't explain succinctly inside a docstring, then alternate styles of documentation like a tutorial or how-to guide would fit better. See this great talk on how to think about organizing documentation, given by one of the guys in the Django project https://www.youtube.com/watch?v=azf6yzuJt54

We really have to use our best judgement on when we can accomplish all of this stuff with a super short docstring or when it's necessary to write more. For now, I'm adding my own personal judgementd in the PR as comments, and hopefully this will give some intuition to anyone reading the so they can also make good choices in the future

**Comment by matentzn** [comment]
Great! Thank you for your input, we are on the same page and I trust your judgement!

File: sssom/cliques.py (hunk 1 above)
Comment (cthoyt) [reply sssom/cliques.py:66]: See #159 for the pull request where I add sphinx-autodoc-typehints, as well as a check to CI that the docs can even be built (currently there is an issue)

Comment (hrshdhgd) [reply sssom/cliques.py:66]: Those are autogenerated by a VSCode plugin ([Python doctoring generator](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring)) that I use for every project. I can take those off if they seem redundant.

Comment (cthoyt) [reply sssom/cliques.py:66]: Not only are they a redundancy, but they're also a liability, since they could get out of sync with the actual type annotations. I'd definitely suggest turning off that part of the docstring generator.

File: sssom/cliques.py [hunk 14]
```diff
@@ -95,7 +108,13 @@ def invert_dict(d: Dict[str, str]) -> Dict[str, str]:
     return invdict
 
 
-def get_src(src, cid):
+def get_src(src: str, cid: str):
```
Comment (cthoyt) [inline sssom/cliques.py:111]: src should be Optional[str] -> see the None check in the definition of the function

File: sssom/sparql_util.py [hunk 15]
```diff
@@ -84,10 +84,22 @@ def query_mappings(config: EndpointConfig) -> MappingSetDataFrame:
...
+
+    :param row: Mapping object row
+    :param config: Configuration
+    :return: Dictionary of CURIEs
+    """
     return {k: contract_uri(v, config) for k, v in row.items()}
 
 
 def contract_uri(uristr: str, config: EndpointConfig) -> str:
+    """Contract URI.
```
Comment (cthoyt) [inline sssom/sparql_util.py:97]: what does it mean to contract a URI

File: sssom/sparql_util.py [hunk 16]
```diff
@@ -84,10 +84,22 @@ def query_mappings(config: EndpointConfig) -> MappingSetDataFrame:
 
 
 def curiefy_row(row: Mapping[str, str], config: EndpointConfig) -> Dict[str, str]:
+    """CURIE-fy row.
```
Comment (cthoyt) [inline sssom/sparql_util.py:87]: not helpful

File: sssom/cli.py [hunk 17]
```diff
@@ -164,7 +164,7 @@ def validate(input: str):
 @main.command()
 @input_argument
 @output_directory_option
-def split(input: str, output_directory: str):
+def split(input: str, output_directory: TextIO):
```
Comment (cthoyt) [inline sssom/cli.py:167]: @hrshdhgd this isn't right - how did you come up with this?

File: sssom/cliques.py [hunk 18]
```diff
@@ -86,16 +95,21 @@ def split_into_cliques(msdf: MappingSetDataFrame):
...
-        if v not in invdict:
-            invdict[v] = []
-        invdict[v].append(k)
-    return invdict
+        rv[v].append(k)
+    return dict(rv)
+
 
+def get_src(src: Optional[str], cid: str):
+    """Get source/CURIE of subject/object in the MappingSetDataFrame.
```
Comment (cthoyt) [inline sssom/cliques.py:107]: this still doesn't make sense by reading it

File: sssom/io.py [hunk 19]
```diff
@@ -81,7 +83,7 @@ def validate_file(input_path: str) -> bool:
         return False
 
 
-def split_file(input_path: str, output_directory: str) -> None:
+def split_file(input_path: str, output_directory: TextIO) -> None:
```
Comment (cthoyt) [inline sssom/io.py:86]: this also isn't correct, why did you change this?

File: sssom/cliques.py [hunk 20]
```diff
@@ -59,7 +60,15 @@ def to_networkx(msdf: MappingSetDataFrame) -> nx.DiGraph:
     return g
 
 
-def split_into_cliques(msdf: MappingSetDataFrame):
+def split_into_cliques(msdf: MappingSetDataFrame) -> List[MappingSetDocument]:
+    """Split MappingSetDataFrames into a list of MappingSetDocuments.
```
Comment (cthoyt) [inline sssom/cliques.py:64]: on what basis? still doesn't describe what a clique is either

File: sssom/context.py [hunk 21]
```diff
@@ -36,6 +52,13 @@ def get_built_in_prefix_map() -> PrefixMap:
 def add_built_in_prefixes_to_prefix_map(
     prefix_map: Optional[PrefixMap] = None,
 ) -> PrefixMap:
+    """Add built-in prefix map from the sssom_context variable in the auto-generated 'internal_context.py' file.
+
+    [Auto generated from sssom.yaml by jsonldcontextgen.py]
+
+    :param prefix_map: Prefix map, defaults to None
```
Comment (cthoyt) [inline sssom/context.py:59]: obviously it defaults to none, that's the default. what happens if None is passed?

File: sssom/parsers.py [hunk 22]
```diff
@@ -632,7 +643,13 @@ def to_mapping_set_document(msdf: MappingSetDataFrame) -> MappingSetDocument:
 
 def split_dataframe(
     msdf: MappingSetDataFrame,
-) -> typing.Mapping[str, MappingSetDataFrame]:
+) -> Dict[str, MappingSetDataFrame]:
+    """Split DataFrame.
```
Comment (cthoyt) [inline sssom/parsers.py:647]: how does it split the dataframe? why?

File: sssom/parsers.py [hunk 23]
```diff
@@ -632,7 +643,13 @@ def to_mapping_set_document(msdf: MappingSetDataFrame) -> MappingSetDocument:
...
+
+    :param msdf: MappingSetDataFrame object
+    :raises RuntimeError: Object is None
```
Comment (cthoyt) [inline sssom/parsers.py:650]: which object?

File: sssom/sparql_util.py [hunk 24]
```diff
@@ -84,10 +84,22 @@ def query_mappings(config: EndpointConfig) -> MappingSetDataFrame:
...
+
+    :param row: Mapping object row
+    :param config: Configuration
+    :return: Dictionary of CURIEs
+    """
     return {k: contract_uri(v, config) for k, v in row.items()}
 
 
 def contract_uri(uristr: str, config: EndpointConfig) -> str:
+    """Contract URI. For e.g. "http://purl.org/dc/terms/" => "dc".
```
Comment (cthoyt) [inline sssom/sparql_util.py:97]: Here's an example of a more informative docstring:

```suggestion
    """Replace the URI with a prefix based on the prefix map in the given configuration.
```

It says what the function does and how it does it. Perhaps consider renaming the variables, too

File: sssom/sparql_util.py [hunk 25]
```diff
@@ -84,10 +84,22 @@ def query_mappings(config: EndpointConfig) -> MappingSetDataFrame:
...
+
+    :param uristr: URI string
+    :param config: Configuration
+    :return: URI string (contracted)
```
Comment (cthoyt) [inline sssom/sparql_util.py:101]: it's not a URI string if it's been contracted. Then it's a prefix.

File: sssom/sparql_util.py [hunk 26]
```diff
@@ -84,10 +84,22 @@ def query_mappings(config: EndpointConfig) -> MappingSetDataFrame:
...
+    :param row: Mapping object row
+    :param config: Configuration
+    :return: Dictionary of CURIEs
+    """
     return {k: contract_uri(v, config) for k, v in row.items()}
 
 
 def contract_uri(uristr: str, config: EndpointConfig) -> str:
+    """Contract URI. For e.g. "http://purl.org/dc/terms/" => "dc".
+
```
Comment (cthoyt) [inline sssom/sparql_util.py:98]: what does it return if it's unable to replace the URI with a prefix? does it error? return none? return the URI again?

File: sssom/sparql_util.py [hunk 27]
```diff
@@ -97,6 +109,12 @@ def contract_uri(uristr: str, config: EndpointConfig) -> str:
 
 
 def expand_curie(curie: str, config: EndpointConfig) -> URIRef:
+    """Expand CURIE. For e.g. "dc" => "http://purl.org/dc/terms/".
```
Comment (cthoyt) [inline sssom/sparql_util.py:112]: What's a CURIE?

Comment (cthoyt) [reply sssom/sparql_util.py:112]: This example isn't demonstrating something on a CURIE

File: sssom/util.py [hunk 28]
```diff
@@ -253,6 +263,11 @@ def filter_redundant_rows(
 
 
 def assign_default_confidence(df: pd.DataFrame) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    """Assign numpy.NaN values to confidence that are blank.
```
Comment (cthoyt) [inline sssom/util.py:266]: use intersphinx linking

```suggestion
Assign :data:`numpy.nan` to confidence that are blank.
```

Comment (cthoyt) [reply sssom/util.py:266]: However, this string doesn't really make sense. The typing says this returns a tuple of two dataframes, and the docstring doesn't say at all what these are or how it's related to assigning default confidence

File: sssom/util.py [hunk 29]
```diff
@@ -348,13 +370,12 @@ def compare_dataframes(df1: pd.DataFrame, df2: pd.DataFrame) -> MappingSetDiff:
...
 
-    Args:
-        df:
-        priors:
-        inverse_factor:
-
-    Returns:
-        List of rows
+    :param df: Pandas DataFrame
+    :param priors: [NOT USED], defaults to None
```
Comment (cthoyt) [inline sssom/util.py:374]: if not used, then delete

File: sssom/util.py [hunk 30]
```diff
@@ -440,13 +461,19 @@ def dataframe_to_ptable(df: pd.DataFrame, priors=None, inverse_factor: float = 0
 RDF_FORMATS = {"ttl", "turtle", "nt", "xml"}
 
 
-def sha256sum(filename):
+def sha256sum(filename: str) -> str:
+    """Hashing function.
```
Comment (cthoyt) [inline sssom/util.py:465]: Here's a more informative, better written docstring:

```suggestion
    """Calculate the SHA256 hash over the bytes in a file.
```

File: sssom/util.py [hunk 31]
```diff
@@ -440,13 +461,19 @@ def dataframe_to_ptable(df: pd.DataFrame, priors=None, inverse_factor: float = 0
...
+
+    :param filename: Filename
```
Comment (cthoyt) [inline sssom/util.py:467]: Would be a good time to rename this variable- `filename` is a bad variable. `path` would be better.

File: sssom/util.py [hunk 32]
```diff
@@ -504,13 +529,10 @@ def merge_msdf(
...
-    taking into account the rule that negative trumps positive given equal confidence values.
-
-    Args:
-        df (pd.DataFrame): Merged Pandas DataFrame
+    Taking into account the rule that negative trumps positive given equal confidence values.
 
-    Returns:
-        pd.DataFrame: Pandas DataFrame with negations addressed
+    :param df: Merged Pandas DataFrame
+    :return: pd.DataFrame: Pandas DataFrame with negations addressed
```
Comment (cthoyt) [inline sssom/util.py:535]: remove type annotation

File: sssom/util.py [hunk 33]
```diff
@@ -458,15 +485,13 @@ def merge_msdf(
...
-        then remove lower confidence positive one. If confidence is the same,
-        prefer HumanCurated. If both HumanCurated, prefer negative mapping).
-
-    Args:
-        msdf1 (MappingSetDataFrame): The primary MappingSetDataFrame
-        msdf2 (MappingSetDataFrame): The secondary MappingSetDataFrame
-        reconcile (bool, optional): [description]. Defaults to True.
+    :param msdf1: The primary MappingSetDataFrame
+    :param msdf2: The secondary MappingSetDataFrame
+    :param reconcile: If reconcile=True, then dedupe(remove redundant lower confidence mappings)
```
Comment (cthoyt) [inline sssom/util.py:490]: break to new line and use tabs, don't justify based on first line (will demonstrate in next commit)

Comment (cthoyt) [reply sssom/util.py:479]: See d298e67

File: sssom/util.py [hunk 34]
```diff
@@ -667,6 +687,12 @@ def inject_metadata_into_df(msdf: MappingSetDataFrame) -> MappingSetDataFrame:
 
 
 def get_file_extension(file: Union[str, TextIO]) -> str:
+    """Get file extension.
+
+    :param file: Name of the file
```
Comment (cthoyt) [inline sssom/util.py:681]: is that really the case? Read the implementation of this function over again.

File: tests/test_data.py [hunk 35]
```diff
@@ -7,28 +8,34 @@
...
-def ensure_test_dir_exists():
-    if not os.path.exists(test_out_dir):
-        os.makedirs(test_out_dir)
-
-
 def load_config():
+    """Load test file information from 'test_config.yaml'.
+
+    :return: Confiuration
+    :rtype: Any
```
Comment (cthoyt) [inline tests/test_data.py:27]: what's the purpose of writing the function returns any?

File: tests/test_data.py [hunk 36]
```diff
@@ -7,28 +8,34 @@
...
 
-def ensure_test_dir_exists():
-    if not os.path.exists(test_out_dir):
-        os.makedirs(test_out_dir)
-
-
 def load_config():
+    """Load test file information from 'test_config.yaml'.
+
+    :return: Confiuration
```
Comment (cthoyt) [inline tests/test_data.py:26]: Typo. Also, if the return is redundant of the one line explanation at the top, you might as well skip it

File: sssom/util.py [hunk 37]
```diff
@@ -267,13 +271,20 @@ def remove_unmatched(df: pd.DataFrame) -> pd.DataFrame:
...
+    :param df: Pandas DataFrame
+    :return: Pandas DataFrame with 'PREDICATE_ID' not 'noMatch'.
     """
     return df[df[PREDICATE_ID] != "noMatch"]
 
 
 def create_entity(row, eid: str, mappings: Dict[str, Any]) -> Entity:
+    """Create an Entity object.
+
+    :param row: Row - [UNUSED]
```
Comment (cthoyt) [inline sssom/util.py:283]: why not just remove it?

File: sssom/util.py [hunk 38]
```diff
@@ -504,13 +518,10 @@ def merge_msdf(
 def deal_with_negation(df: pd.DataFrame) -> pd.DataFrame:
     """Combine negative and positive rows with matching [SUBJECT_ID, OBJECT_ID, CONFIDENCE] combination.
 
-    taking into account the rule that negative trumps positive given equal confidence values.
+    Taking into account the rule that negative trumps positive given equal confidence values.
```
Comment (cthoyt) [inline sssom/util.py:521]: not a sentence. please fix grammar

File: sssom/util.py [hunk 39]
```diff
@@ -679,7 +694,14 @@ def get_file_extension(file: Union[str, TextIO]) -> str:
         raise Exception(f"Cannot guess format from {filename}")
 
 
-def read_csv(filename, comment="#", sep=","):
+def read_csv(filename: str, comment: str = "#", sep: str = ",") -> pd.DataFrame:
+    """Read TSV file.
```
Comment (cthoyt) [inline sssom/util.py:698]: fix grammar

File: sssom/util.py [hunk 40]
```diff
@@ -721,10 +743,15 @@ def read_pandas(file: Union[str, TextIO], sep: Optional[str] = None) -> pd.DataF
         else:
             sep = "\t"
             logging.warning("Cannot automatically determine table format, trying tsv.")
-    return read_csv(file, comment="#", sep=sep).fillna("")
+    return read_csv(str(file), comment="#", sep=sep).fillna("")
```
Comment (cthoyt) [inline sssom/util.py:746]: this str() is highly suspect. what if a file object was passed?

File: sssom/util.py [hunk 41]
```diff
@@ -770,10 +799,20 @@ class NoCURIEException(ValueError):
...
+    """Check if string is CURIE or not.
+
+    :param string: String
+    :return: Boolean indicating CURIE or not
+    """
     return bool(CURIE_RE.match(string))
 
 
 def get_prefix_from_curie(curie: str) -> str:
+    """Get prefix from CURIE.
```
Comment (cthoyt) [inline sssom/util.py:811]: fix grammar

File: sssom/util.py [hunk 42]
```diff
@@ -770,10 +799,20 @@ class NoCURIEException(ValueError):
 
 
 def is_curie(string: str) -> bool:
+    """Check if string is CURIE or not.
```
Comment (cthoyt) [inline sssom/util.py:802]: fix grammar

File: sssom/util.py [hunk 43]
```diff
@@ -812,14 +851,25 @@ def curie_from_uri(uri: str, prefix_map: Mapping[str, str]) -> str:
 
 
 def get_prefixes_used_in_table(df: pd.DataFrame) -> List[str]:
+    """Get prefixes used in table.
+
+    :param df: Pandas DatafFrame
+    :return: List of unique prefixes
```
Comment (cthoyt) [inline sssom/util.py:857]: this function is simple enough that the param and return values are redundant. just use a single line

File: sssom/util.py [hunk 44]
```diff
@@ -812,14 +851,25 @@ def curie_from_uri(uri: str, prefix_map: Mapping[str, str]) -> str:
 
 
 def get_prefixes_used_in_table(df: pd.DataFrame) -> List[str]:
+    """Get prefixes used in table.
```
Comment (cthoyt) [inline sssom/util.py:854]: should write dataframe and not table. which parts of the dataframe does it get prefixes from?

File: sssom/util.py [hunk 45]
```diff
@@ -812,14 +851,25 @@ def curie_from_uri(uri: str, prefix_map: Mapping[str, str]) -> str:
...
     prefixes = []
     for col in KEY_FEATURES:
         for v in df[col].values:
             prefixes.append(get_prefix_from_curie(v))
     return list(set(prefixes))
 
 
-def filter_out_prefixes(df: pd.DataFrame, filter_prefixes) -> pd.DataFrame:
+def filter_out_prefixes(df: pd.DataFrame, filter_prefixes: List[str]) -> pd.DataFrame:
+    """Get all prefixes in filter_prefixes from pandas DataFrame.
```
Comment (cthoyt) [inline sssom/util.py:867]: this docstring does not match what the function does

File: sssom/util.py [hunk 46]
```diff
@@ -836,6 +886,12 @@ def filter_out_prefixes(df: pd.DataFrame, filter_prefixes) -> pd.DataFrame:
 
 
 def guess_file_format(filename: Union[str, TextIO]) -> str:
+    """Get file format.
```
Comment (cthoyt) [inline sssom/util.py:889]: get or guess? need better docstring that gives more insight.

File: sssom/io.py (hunk 19 above)
Comment (hrshdhgd) [reply sssom/io.py]: mypy was expecting a TextIO parameter instead of a str. So I switched it which made the error go away. I ran pytest immediately after to see it broke anything and the tests passed. So I stuck with it.

File: sssom/cliques.py (hunk 9 above)
Comment (cthoyt) [reply sssom/cliques.py]: 6fd4230

**Comment by matentzn** [comment]
Just a small executive decision here: We will address all the comments currently in the PR, but then we need to take a break from refactoring and documenting.. The todolist of feature requests is piling up, and I need to weigh the benefits of retrofitting docs right now to the importance of moving on. So @hrshdhgd, you can give this PR a last pass, and @matentzn me anywhere you prefer me to do it. :) Happy to! Thank you both for your efforts!

File: sssom/context.py (hunk 21 above)
Comment (hrshdhgd) [reply sssom/context.py:59]: @matentzn

File: sssom/parsers.py (hunk 22 above)
Comment (hrshdhgd) [reply sssom/parsers.py:647]: @matentzn

File: sssom/sparql_util.py (hunk 26 above)
Comment (hrshdhgd) [reply sssom/sparql_util.py:98]: @matentzn

File: sssom/util.py (hunk 37 above)
Comment (hrshdhgd) [reply sssom/util.py:283]: @matentzn

File: sssom/util.py (hunk 29 above)
Comment (hrshdhgd) [reply sssom/util.py:363]: @matentzn

**Comment by hrshdhgd** [comment]
flake8 passes mypy gives 2 errors. @cthoyt, It is a domino effect of changing the annotation type at one place and trickle down to other places which may result several `Union[]`s . Could you please provide some guidance? Thanks in advance!

**Comment by cthoyt** [comment]
> flake8 passes mypy gives 2 errors. @cthoyt, It is a domino effect of changing the annotation type at one place and it impacts other places which may result several `Union[]`s . Could you please provide some guidance? Thanks in advance!

Just keep in mind it's not just about passing CI, but also about doing documentation and style well. 

The cascade of type checking is actually a feature - if many things rely on each other and become out of sync, it's good that mypy fails. I can't give any general guidance other than that when this happens, you should really do some digging to figure out where the type annotations went wrong - read all of the other functions that seemed to break

File: sssom/util.py [hunk 47]
```diff
@@ -743,7 +716,7 @@ def read_pandas(file: Union[str, TextIO], sep: Optional[str] = None) -> pd.DataF
         else:
             sep = "\t"
             logging.warning("Cannot automatically determine table format, trying tsv.")
-    return read_csv(file, comment="#", sep=sep).fillna("")
+    return pd.read_csv(file, comment="#", sep=sep).fillna("")
```
Comment (hrshdhgd) [inline sssom/util.py:711]: The reason why `read_csv` was created was due to pandas omitting rows within the `DataFrame` of a `MappingSetDataFrame` object which would be part of the data and not a comment. @matentzn and I had discussed this while implementing the function.

File: sssom/sparql_util.py (hunk 26 above)
Comment (cthoyt) [reply sssom/sparql_util.py:98]: These aren't really philosophical questions, I'd suggest reading the code itself and reasoning about what it does directly. The answer seems pretty obvious from my side, but I think it's a good exercise for you to try yourself too

File: sssom/util.py (hunk 47 above)
Comment (cthoyt) [reply sssom/util.py:711]: was that because the data had the comment character (`#`) in it?

File: sssom/util.py (hunk 37 above)
Comment (matentzn) [reply sssom/util.py:279]: as discussed in call, just remove all mentions of row. MAke sure all calls to create_entity use parameter names (create_entity(eid="a"), not create_entity(eid))

File: sssom/util.py (hunk 47 above)
Comment (hrshdhgd) [reply sssom/util.py:708]: Yes.

Comment (matentzn) [reply sssom/util.py:708]: Yes this looks like a very dangerous change

Comment (cthoyt) [reply sssom/util.py:708]: okay then i'll revert it.

Comment (cthoyt) [reply sssom/util.py]: Reverted in cf80b47

File: sssom/util.py (hunk 37 above)
Comment (cthoyt) [reply sssom/util.py:281]: looks like this is done

File: sssom/util.py (hunk 45 above)
Comment (cthoyt) [reply sssom/util.py]: fixed in 0f74c64

**Comment by cthoyt** [comment]
Okay guys, thanks so far for the monumental effort on this PR. Everything is passing CI, and I've gone through and made tweaks to everything that had a comment on it (importantly including reverting that change I made on `read_csv()`). Have a last read-through at the beginning of the week and let's get this one merged :)

**Review by matentzn — APPROVED** [review]
I am amazed. Thank you both for this work! Really made my day :)
